packages/automation-extra/src/base.ts

```diff
@@ -0,0 +1,335 @@
+    return browser
+  }
+
+  protected async _bindPuppeteerBrowserEvents(browser: pptr.Browser) {
+    debug('_bindPuppeteerBrowserEvents')
+
+    browser.on('disconnected', () => {
+      this.plugins.dispatch('onDisconnected', browser)
+      this.plugins.dispatchLegacy('onClose')
+    })
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    browser.on('targetcreated', async (target: pptr.Target) => {
+      debug('targetcreated')
+      this.plugins.dispatchLegacy('onTargetCreated', target)
+      // Pre filter pages for plugin developers convenience
+      if (target.type() === 'page') {
+        const page = await target.page()
+        page.on('close', () => {
+          this.plugins.dispatch('onPageClose', page)
+        })
+
+        this.plugins.dispatch('onPageCreated', page)
+      }
+    })
+    // // Legacy events
+    browser.on('targetchanged', (target: pptr.Target) => {
+      this.plugins.dispatchLegacy('onTargetChanged', target)
+    })
+    browser.on('targetdestroyed', (target: pptr.Target) => {
+      this.plugins.dispatchLegacy('onTargetDestroyed', target)
+    })
+  }
+
+  protected async _bindPlaywrightBrowserEvents(browser: pw.Browser) {
+    debug('_bindPlaywrightBrowserEvents')
+
+    browser.on('disconnected', () => {
+      this.plugins.dispatch('onDisconnected', browser)
+    })
+
+    const bindContextEvents = (context: pw.BrowserContext) => {
+      // Make sure things like `addInitScript` show an effect on the very first page as well
+      context.newPage = ((originalMethod, ctx) => {
+        return async () => {
+          const page = await originalMethod.call(ctx)
+          await page.goto('about:blank')
+          return page
+        }
+      })(context.newPage, context)
+
+      context.on('close', () => {
+        this.plugins.dispatch('onContextClose', context)
+      })
+      context.on('page', page => {
+        this.plugins.dispatch('onPageCreated', page)
+
+        page.on('close', () => {
+          this.plugins.dispatch('onPageClose', page)
+        })
+      })
+    }
+
+    browser.newContext = ((originalMethod, ctx) => {
+      return async (options: pw.BrowserContextOptions = {}) => {
+        const contextOptions: pw.BrowserContextOptions =
+          (await this.plugins.dispatchBlocking(
+            'beforeContext',
+            options || {},
+            browser
+          )) || options
+        const context = await originalMethod.call(ctx, contextOptions)
+        this.plugins.dispatch('onContextCreated', context, contextOptions)
+
+        bindContextEvents(context)
+        return context
+      }
+    })(browser.newContext, browser)
+
+    browser.newPage = ((originalMethod, ctx) => {
+      return async (options: pw.BrowserContextOptions = {}) => {
+        const contextOptions: pw.BrowserContextOptions =
+          (await this.plugins.dispatchBlocking(
+            'beforeContext',
+            options,
+            browser
+          )) || options
```

Comment:
I believe this dispatch call is a bug. Playwright's `browser.newPage()` internally calls `browser.newContext()` then `context.newPage()` so this is firing the `beforeContext` event twice for every `browser.newPage()` call. [Here's a link to the relevant file & line number in Playwright.](https://github.com/microsoft/playwright/blob/master/src/client/browser.ts#L67)

Reply:
Overall, this is an excellent refactor! I have been using this branch quite a lot while working with Playwright and just started noticing a few bugs here and there.

Reply:
Very interesting, thanks! Will do some local testing around the `beforeContext` dispatch, it's possible for the second call to be redundant.

> Overall, this is an excellent refactor! I have been using this branch quite a lot while working with Playwright and just started noticing a few bugs here and there.

As you're the first brave alpha user of `automation-extra` (😅) kindly mention any bug you might encounter here or on discord, thanks :-)

Reply:
@alec-mccormick You were correct, the event was fired twice. :-) The tests didn't catch that as they only checked if the event was fired, not how often. Fix & updated tests: [`4bad9e6` (#303)](https://github.com/berstend/puppeteer-extra/pull/303/commits/4bad9e6daca6cfdede5de230e37c84b747512f33)

Thanks!